### PR TITLE
Implementation of robust publication

### PR DESCRIPTION
**Pull request by cnweaver**
## Description

This patch is not entirely complete and ready to merge, but I would like to begin reviewing it for reasons discussed below. All naming and organization should be up for discussion; I wasn't sure where we would want things so I have put them in a new file and given them 'strawman' names. Since we may want to change some or all of this, the usage documentation is not yet written; obviously that must be done before the review and merge complete. 

As a specific, more minor issue, the linter is not happy with `hop.robust_publisher.PublicationJournal._read_previous_journal`. The 
method is indeed fairly long, but most of my ideas for splitting it up would add more conceptual complexity, as state that currently exists shared within it would have to be shared among multiple methods. Suggestions would be welcome. 

## Checklist

* [x] All new functions and classes are documented and adhere to Google doc style (3.8.3-3.8.6 of [this](http://google.github.io/styleguide/pyguide.html#383-functions-and-methods) document)
* [x] Add/update sphinx documentation with any relevant changes.
* [x] Add/update pytest-style tests in `/tests`, ensuring sufficient code coverage.
* [x] `make test` runs without errors.
* [x] `make lint` doesn't give any warnings.
* [x] `make format` doesn't give any code formatting suggestions.
* [x] `make doc` runs without errors and generated docs render correctly.
* [x] Check that CI pipeline run on this PR passes all stages.
* [ ] Review signoff by at least one developer.

### CODE REVIEW

**Review by bfc5288 — COMMENTED** [review]
Hi @cnweaver -- I've been making my first full pass at this and wanted to send my first set of comments (they are mostly basic points or clarifications as I'm still learning the structure of the new code). I committed a few fixes to typos.

I will add more comments as I continue the review.

File: hop/robust_publisher.py [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+        """
+        if sequence_number not in self.maybe_sent_messages:
+            raise RuntimeError("No record of message with sequence number " + str(sequence_number)
+                               + " being queued for sending")
+        self._write_record(PublicationJournal.sent_record_type,
+                           PublicationJournal.encode_int(sequence_number))
+        del self.maybe_sent_messages[sequence_number]
+        if len(self.messages_to_send) == 0 and len(self.maybe_sent_messages) == 0:
+            # take this opportunity to garbage collect the journal file
+            self.journal.close()
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply hop/robust_publisher.py:261]: The way I had thought about it was that the journal should be kept around unless we can prove we don't need I anymore. In general, one would like to garbage collect all pairs of matching (message-to-send, message-sent) records, but keeping the format simple and append-only meant that this can only be done when all queued messages are known to have been sent. It might be possible to do some somehwat fancier things (let the journal split into multiple files, implement it as a riing buffer, etc.) which would allow finer grained clean up, or clean up at other times, but off the top of my head, when we know we've sent a message is the only state change in the current scheme when we know we can clean up.

File: hop/robust_publisher.py [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+        Args:
+            journal_path: The filesystem path from/to which the journal data should be read/written.
+
+        Raises:
+            PermissionError: If existing journal file does not have suitable permissions.
+            RuntimeError: If existing journal data cannot be read.
+        """
+        self.journal_path = journal_path
+        # each queue will store (message, headers) tuples, keyed by sequence numbers
+        self.messages_to_send = RAPriorityQueue()
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply hop/robust_publisher.py:141]: This seems less clear to me since it would then not mention the purpose of the queue (as there are two queues, currently `messages_to_send` and `messages_to_send`)?

File: tests/test_rpub.py [hunk 3]
```diff
@@ -0,0 +1,837 @@
...
+        raise OSError(errno.EIO, os.strerror(errno.EIO))
+    file_mock = MagicMock()
+    file_mock.write = write_fail
+    file_mock.__enter__ = MagicMock(return_value=file_mock)
+    orig_open = open
+    # allow reading to proceed normally so we can read the existing journal,
+    # but intercept writing to inject failures
+
+    def sneaky_open(path, mode, *args, **kwargs):
+        if 'r' in mode:
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply tests/test_rpub.py]: Done

File: tests/test_rpub.py [hunk 4]
```diff
@@ -0,0 +1,837 @@
...
+    j = PublicationJournal(journal_path)  # relaod with message to send
+    assert j.has_messages_to_send()
+    assert not j.has_messages_in_flight()
+
+    s0 = j.get_next_message_to_send()
+    assert s0[1] == m0, "Message body should be preserved by on-disk journal"
+    assert s0[2] == headers, "Message headers should be preserved by on-disk journal"
+    del j
+    j = PublicationJournal(journal_path)  # relaod with message in flight
+    # this may appear counter-intuitive, but if the journal/sender is offline while the message is
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply tests/test_rpub.py:319]: Good point, I have added a paragraph about this.

File: hop/robust_publisher.py [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+    def __len__(self):
+        """Return: the number of items in the queue."""
+        assert len(self.data) == len(self.priorities)
+        return len(self.data)
+
+    def __contains__(self, key):
+        """Test whether an item with the given key/priority is present in the queue."""
+        return key in self.data
+
+    def insert(self, key, value):
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply hop/robust_publisher.py:40]: This is kind of horrifying, and seems to fundamentally be a bug in `heapq.heappush`, which does not seem to meet even the basic exception safety guarantee. If `heapq.heappush` were safe, I think this could be dealt with on my end by doing the insertion into `priorities` (which might fail), before the insertion into `data`, but that's unfortunately not viable for the immediate future. 

Things I can see reasonably simply doing about this would be:

1) A documentation-only fix which just tells the user not to mix non-orderable key types. This would work for our immediate purposes, since `PublicationJournal` only ever uses a single key type anyway. 

2) Strong typing with a runtime check: I could make the queue constructor take a type argument for the expected key type, and enforce that all supplied keys are that type. This is probably dreadfully 'unpythonic', and also the only way I can think of to make it actually safe.

File: tests/test_rpub.py [hunk 6]
```diff
@@ -0,0 +1,837 @@
...
+########################################
+
+
+def test_journal_construct_no_file(tmpdir):
+    journal_path = tmpdir.join("journal")
+    j = PublicationJournal(journal_path)
+    assert not j.has_messages_to_send()
+    assert not j.has_messages_in_flight()
+    assert j.get_next_message_to_send() is None
+    assert os.path.exists(journal_path)
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply tests/test_rpub.py:144]: The `PublicationJournal` constructor creates the journal file, so the check is correct, if not very useful/important. It could be argued that this isn't necessary, so we shouldn't test for it, but I would not want to enforce the reverse, as it would complicate the journal implementation to have to open the journal for writing lazily.

File: tests/test_rpub.py [hunk 7]
```diff
@@ -0,0 +1,837 @@
...
+    # directly use internal recording function to generate records with valid checksums
+    # but dubious meanings
+    test_message = b"data"
+    rbody = BytesIO()
+    rbody.write(PublicationJournal.encode_int(0))  # sequence number
+    rbody.write(PublicationJournal.encode_int(len(test_message)))  # data length
+    rbody.write(test_message)  # data
+    rbody.write(PublicationJournal.encode_int(0))  # no headers
+    j._write_record(PublicationJournal.msg_record_type, rbody.getvalue())
+    # record the same message again, duplicating the sequence number
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply tests/test_rpub.py:444]: True, and that was more generally what the test was intended for, since any duplicate sequence number is invalid. Originally I was being lazy and not encoding a seocnd record, but I've now extended the test to do that, too.

File: hop/robust_publisher.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bfc5288) [reply hop/robust_publisher.py:40]: I think that (1) would be reasonable as long as this remains a bug (rather than an exploit). But feel free to try (2) if it is that straightforward.

File: hop/robust_publisher.py [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+                do_send = False
+                if journal.has_messages_to_send():
+                    seq_num, message, headers = journal.get_next_message_to_send()
+                    do_send = True
+            if do_send:
+                try:
+                    dc = journal.get_delivery_callback(seq_num, self._lock)
+                    logger.debug(f"Sending message with sequence number {seq_num}")
+                    self._stream.write_raw(message, headers=headers, delivery_callback=dc)
+                except KafkaException as e:
```
Comment (bfc5288) [inline hop/robust_publisher.py:612]: This appears to be redundant with the `get_delivery_callback` call a few lines above, which would also lead to a logger.error and message requeue (via `_delivery_callback`). Is this desired?

**Review by bfc5288 — COMMENTED** [review]
Here are a few minor comments. I am now looking into the threading/locking structure to check for issues.

File: tests/test_rpub.py [hunk 9]
```diff
@@ -0,0 +1,854 @@
...
+
+    # directly use internal recording function to generate records with valid checksums
+    # but dubious meanings
+    # record that a message was sent without it being previously mentioned
+    j._write_record(PublicationJournal.sent_record_type, PublicationJournal.encode_int(56))
+    del j
+
+    with pytest.raises(RuntimeError) as excinfo:
+        j = PublicationJournal(journal_path)
+    assert "Record of sent message" in str(excinfo.value)
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply tests/test_rpub.py:518]: I think that's a different type of error, and is covered by the existing `test_journal_restore_too_short_message_record`. That test uses `PublicationJournal.msg_record_type`, and writes an invalid message for that type (which, among other issues, is too short). `test_journal_restore_sent_unknown_sequence_number` is using `PublicationJournal.sent_record_type`, which has a much simpler corresponding layout, which really is supposed to be just a single encoded integer. So, the test here is writing a record which is internally valid, but while is logically nonsensical when the previous message/record it references does not exist.

File: hop/robust_publisher.py [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+                    raise RuntimeError(f"Journal corrupted: Claimed record body length ({rec_len}) "
+                                       f"too small to conain required data for record "
+                                       f"{record_index} which began at file offset {record_start}")
+
+                raw_seq_num = read(self.int_size, "message sequence number")
+                seq_num, bcrc = decode(raw_seq_num, decode_int, bcrc, "sequence number")
+
+                raw_msg_len = read(self.int_size, "record message length")
+                msg_len, bcrc = decode(raw_msg_len, decode_int, bcrc, "message length")
+                if msg_len > (rec_len - required_body_len):
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply hop/robust_publisher.py:358]: Sure; I've added a test for this.

File: hop/robust_publisher.py [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+                    logger.debug(f"Sending message with sequence number {seq_num}")
+                    self._stream.write_raw(message, headers=headers, delivery_callback=dc)
+                except KafkaException as e:
+                    logger.error(f"Error sending message with sequence number: {seq_num}: {e}"
+                                 "; requeuing to send again")
+                    with self._lock:
+                        journal.requeue_message(seq_num)
+            try:
+                self._stream._producer._producer.poll(self._poll_wait)
+            except KafkaException:
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply hop/robust_publisher.py:619]: This isn't documented, but callbacks crashing or raising python errors seem to be possible causes. It's possible that an older version of the callback stuff required catching and squashing errors here; unfortunately I'm just not sure of that or whether it's still needed.

File: hop/robust_publisher.py (hunk 8 above)
Comment (cnweaver) [reply hop/robust_publisher.py:612]: I think that the unedrlying confluent-kafka-python function can fail in a number of ways before it could invoke a callback, for example if it has an argument parsing error, or for certain types of internal version mismatches in the C libraries.

File: hop/robust_publisher.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bfc5288) [reply hop/robust_publisher.py:619]: Could this be documented somewhere just in case?

**Review by bfc5288 — COMMENTED** [review]
I have been looking into the threading and locking mechanisms and I think things look good overall.

The last thing I wanted to clarify is under what circumstances the RP would be started. E.g., will it always be used as a context manager? If not, there could be a situation in which an RP is initialized and `run()` is called but it stays in the `_do_send` loop:
```
rp = RobustPublisher(url,auth=False)
rp.run()
# hanging in _cond.wait()
```
... or is initialized and writes messages but those messages are not sent since `run()` isn't called, which also leads to hanging:
```
rp = RobustPublisher(url,auth=False)
for i in range(1,10): msg = loader.load(f'test message {i}'); rp.write(msg); time.sleep(0.1)
# no messages are sent
rp.run()
# hanging
^C

del rp
rp = RobustPublisher(url,auth=False)
rp.run()
# messages are sent
```

N.b., I'm not sure if either of these cases would apply in actual usage, so this may not matter. If not, then I think we can move on and integrate it with the actual publisher, add documentation, etc.

The last item is to make sure we add documentation about what key types should be used ([original discussion here](https://github.com/scimma/hop-client/pull/174/files#r783156338)).

**Comment by cnweaver** [comment]
> The last thing I wanted to clarify is under what circumstances the RP would be started. E.g., will it always be used as a context manager?

As it is currently written, it really should only be used as a context manager, since there's no way to correctly stop it besides calling `RobustPublisher.__exit__`. In particular, `RobustPublisher.run()` is not intended to ever be called from the main thread, or indeed directly by the user at all (it should be called by `start` doing whatever magic it does). I would really prefer to make this function private, or since python doesn't support that, at least mark it as internal with an underscored name, but that wouln't conform to the interface required by `threading.Thread`. I think this is probably a case where the best that can be done is putting a big warning in the documentation telling users not to do call it directly. 

If we want to support use not as a context manager, I think we can do that simply enough. Calling `start` is the right way to start the background thread, and implicitly call `run`, and is all `__enter__` does. `__exit__` does a bit more, but that could be factorred out into a `stop` function which users can call if they want to control the object manually. I do think use as a context manager should be the recommended way, but giving power users a bit more flexibility to do things I may not have thought of seems fine, too. 

> I'm not sure if either of these cases would apply in actual usage, so this may not matter. 

As things stand, it seems very easy for a user to make the mistake of calling `run` directly from the main thread and then being deadlocked, so I think at least documenting that that should not be done is important. 

> The last item is to make sure we add documentation about what key types should be used

Sorry, I did that but forgot to push the commit. It should be taken care of in c90d43197d918c640cf80c7817ea66c11ab19a4d.

**Comment by bfc5288** [comment]
> I would really prefer to make this function private, or since python doesn't support that, at least mark it as internal with an underscored name, but that wouln't conform to the interface required by `threading.Thread`. I think this is probably a case where the best that can be done is putting a big warning in the documentation telling users not to do call it directly.

This makes sense. I am a bit surprised that there isn't a way to make this method private, given the prevalence of threading. Without a more robust alternative, documentation would be fine.

> If we want to support use not as a context manager, I think we can do that simply enough. Calling `start` is the right way to start the background thread, and implicitly call `run`, and is all `__enter__` does. `__exit__` does a bit more, but that could be factorred out into a `stop` function which users can call if they want to control the object manually. I do think use as a context manager should be the recommended way, but giving power users a bit more flexibility to do things I may not have thought of seems fine, too.

This could be good to support, though using it as a context manager is probably the better option. And I am not sure of the potential use cases that would benefit from a non-context-managed approach, so I am fine with coming back to restructure this at a later point if it ends up being a limitation.

File: hop/robust_publisher.py [hunk 12]
```diff
@@ -0,0 +1,640 @@
...
+                        "left-over messages to send")
+
+        self._lock = threading.Lock()
+        self._cond = threading.Condition(self._lock)
+
+        dummy = io.Stream(auth=auth)
+        self._stream = dummy.open(url, "w", error_callback=PublicationJournal.error_callback,
+                                  **kwargs)
+
+        self._running = True
```
[the remark this answers is not in the document]
Comment (cnweaver) [reply hop/robust_publisher.py]: I looked at this again, and renamed the variable `_should_stop`, since that is what it is really used to indicate, and is more along the lines of how I usually write this pattern.

**Comment by cnweaver** [comment]
> This could be good to support, though using it as a context manager is probably the better option. 

This is now included, with some basic, inline documentation of the requirements relating the various methods.

**Comment by bfc5288** [comment]
This looks good. For implementation, how should this be integrated with the regular publisher, i.e. as an CLI option, only via the API, etc.? It may make sense to have it be enabled by default, though it is conceivable that there may be users who don't want `hop-client` to write anything to disk.

**Comment by cnweaver** [comment]
I was leaning towards having it be a python API-only feature? As you say, CLI users might not expect things to be written to disk, etc., and I view the CLI as being great for quick interactive use, but probably not what we want serious users to rely on for important purposes. However, integrating it into the CLI is entirely doable if people think it is worthwhile, and we could then also consider whether it should be the default publication mode.

**Comment by bfc5288** [comment]
> As a specific, more minor issue, the linter is not happy with `hop.robust_publisher.PublicationJournal._read_previous_journal`. The method is indeed fairly long, but most of my ideas for splitting it up would add more conceptual complexity, as state that currently exists shared within it would have to be shared among multiple methods. Suggestions would be welcome.

I think that `hop.robust_publisher.PublicationJournal._read_previous_journal.read` could be moved out; I believe the only adjustment this would entail is passing a `journal` object (or just `buffer`) to `read`. This reduces the flake8 complexity metric from 26 to 20 (the threshold for our linting is 15).

Otherwise, I am not sure a reasonable way to reduce the complexity metric while also reducing the actual complexity of the code. One idea would be to move out most of the `rec_type == ...` record checking logic to a separate function like _check_journal_records(journal, bcrc,...) which would include this stanza ([L348](https://github.com/scimma/hop-client/blob/8b297a363e4b3832fd15835643097b206cf3d9dd/hop/robust_publisher.py#L348) through [L442](https://github.com/scimma/hop-client/blob/8b297a363e4b3832fd15835643097b206cf3d9dd/hop/robust_publisher.py#L442)):
```
            if rec_type == PublicationJournal.msg_record_type:
            ....
            else:
                raise RuntimeError(f"Journal corrupted: Invalid record type ({rec_type}) "
                                   f"at file offset {record_type_offset}")
```
Another alternative could be to somehow separate only the first `rec_type == PublicationJournal.msg_record_type` check, which would lower the complexity metric enough. I am not sure if either of these two approaches would lower the actual complexity, but they would bring the metric within the threshold.

**Comment by cnweaver** [comment]
I've moved stuff around (`read` as suggested, and also the header reading logic), which triggered having to factor out several other smaller pieces. The complexity is definitely worse overall now, with local helper functions which have no general use now appearing at the class scope, and large amounts of local state getting handed back and forth, but the linter is appeased, as its very narrow understanding of complexity (basically branches per function?) has been reduced.

**Comment by cnweaver** [comment]
I already did the rectoring, so I don't think I have any strong desire to ask for a linting exeception in this case.

**Review by bfc5288 — APPROVED** [review]
This looks goods to me. Great work. Feel free to merge when convenient.
